src/ctia/task/migration/migrate_es_stores.clj

```diff
@@ -104,59 +185,28 @@
-                                                   batch-size
-                                                   0
-                                                   "asc"
-                                                   current-search-after)
-            next (:next paging)
-            next-search-after (:sort paging)]
-        (>!! data-chan {:documents data
-                        :search_after next-search-after})
-        (if next
-          (recur  next-search-after)
-          (close! data-chan))))
-
-    (loop [migrated-count migrated-count-state]
-      ;; sync loop that consumes batches in chan and migrates them to target
-      (let [{:keys [documents search_after]} (<!! (go (<! data-chan)))
-            migrated (transduce migrations conj documents)
-            {:keys [data errors]} (list-coerce migrated)
-            new-migrated-count (+ migrated-count (count data))]
-        (doseq [entity errors]
-          (let [message
-                (format "%s - Cannot migrate entity: %s"
-                        (:type source-store)
-                        (pr-str entity))]
-            (log/error message)))
-        (when confirm?
-          (when (seq data) (mst/store-batch target-store data))
-          (mst/rollover target-store batch-size new-migrated-count)
-          (mst/update-migration-store migration-id
-                                      entity-type
-                                      (into {:target {:migrated new-migrated-count}}
-                                            (when search_after
-                                              {:source {:search_after search_after}}))
-                                      @mst/migration-es-conn))
-        (log/infof "%s - migrated: %s documents"
-                   (:type source-store)
-                   new-migrated-count)
-        (swap! batches dec)
-        (if (< 0 @batches)
-          (recur new-migrated-count)
-          (do (log/infof "%s - finished migrating %s documents"
-                         (:type source-store)
-                         new-migrated-count)
-              (when confirm?
-                (mst/finalize-migration! migration-id
-                                         entity-type
-                                         source-store
-                                         target-store
-                                         @mst/migration-es-conn))))))))
+    (loop [queries all-queries
+           migrated-count migrated-count-state]
+      (if-let [query (first queries)]
+        (let [migration-params (assoc base-params
+                                      :query query
+                                      :migrated-count migrated-count
+                                      :data-chan (chan buffer-size))]
+          (log/infof "%s - handling sliced query %s"
+                     (name entity-type)
+                     (pr-str query))
+          (thread (read-source migration-params))
```

Comment:
here if I'm not mistaken you treat each query sequentially,
you first `read-source` then finally  you `write-target` on `recur`.
there is no need for the `thread` since you do it one by one and in this regard you could totally pass an atom instead of a chan or even use `write source` passing the result of read-source as a var.

after investigating the process I believe your approach is valid though but if we could remove an unnecessary threading logic from the loop that would help understand what's happening at a glance.